### PR TITLE
Introduce Pickles drawio diagram

### PR DESCRIPTION
**Pull request by volhovm**
This introduces the pickles structure diagram, intended to be a part of the mina book.

### CODE REVIEW

File: book/src/pickles/diagrams.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Pickles Technical Diagrams
+
+This section contains a series of diagrams giving an overview of implementation of pickles, closely following the structures and abstractions in the code. However, they are very technical and are primarily intended for developer use. The primary source of all the images in this section is [`pickles_structure.drawio`](./pickles_structure.drawio) file, and if one edits it one must re-generate the `.svg` renders by doing "export -> selection".
+
+The legend of the diagrams is quite straightforward:
+- The black boxes are data structures that have names and labels following the implementation.
+  - `MFNStep`/`MFNWrap` is an abbreviation from `MessagesForNextStep` and `MessagesForNextWrap` that is used for brewity. Most other datatypes are exactly the same as in the codebase.
```
Comment (mrmr1993) [inline book/src/pickles/diagrams.md:7]: `s/brewity/brevity`

File: book/src/pickles/overview.md [hunk 2]
```diff
@@ -1,14 +1,30 @@
...
-3. Checking
-
-    - `RecursionChallenge.comm` $\in \GG$ fields from each "input proof"
-    - The polynomial commitments in the
-    - $u \in [0, 2^{128})$ challenge (provided in GLV form).
-3. Compute $y$ from:
-    - `RecursionChallenge.prev_challenges` $\in \FF^\ell$ fields from each "input proof"
+# Overview of Pickles
+
+Pickles is a recursion layer built on top of Kimchi. The complexity of pickles as a protocol lies in specifying how to verify previous kimchi inside of the current ones. And it gets quite complicated with many concrete details. Working over two curves requires us to have different circuits and a "mirrored" structure, some computations are deferred for efficiency, and one needs to carefully keep track of the accumulators. In this section we provide a general overview of pickles, while next sections in the same chapter dive into the actual implementation details.
```
Comment (mrmr1993) [inline book/src/pickles/overview.md:3]: Delete `And it gets quite complicated with many concrete details.`

File: book/src/pickles/overview.md [hunk 3]
```diff
@@ -1,14 +1,30 @@
...
+
+Pickles works over [Pasta](/specs/pasta.md), a cycle of curves consisting of Pallas and Vesta, and thus it defines two generic circuits, one for each curve. Each can be thought of as a parallel instantiation of a kimchi proof systems. These circuits are not symmetric and have somewhat different function:
+- **Step circuit**: this is the main circuit that contains application logic. Each step circuit verifies a statement and potentially several (at most 2) other wrap proofs.
+- **Wrap circuit**: this circuit merely verifies the step circuit, and does not have its own application logic. The intuition is that every time an application statement is proven it's done in Step, and then the resulting proof is immediately wrapped using Wrap.
+
+
+#### General Circuit Structure
+
+Both Step and Wrap circuits additionally do a lot of recursive verification of the previous steps. Without getting too technical, Step (without lost of generality) does the following:
+1. Verify the application logic statement (e.g. the mina transaction is valid)
```
Comment (mrmr1993) [inline book/src/pickles/overview.md:13]: `Verify` here feels misleading. It's more `Prove` or `Execute`

File: book/src/pickles/overview.md [hunk 4]
```diff
@@ -1,14 +1,30 @@
...
+2. Verify that the previous Wrap proof is (first-)half-valid (perform only main checks that are efficient for the curve)
+3. Verify that the previous Step proof is (second-)half-valid (perform the secondary checks that were inefficient to perform when the previous Step was Wrapped)
+4. Verify that the previous Step correctly aggregated the previous accumulator, e.g. $\mathsf{acc}_2 = \mathsf{Aggregate}(\mathsf{acc}_1, \pi_{\mathsf{step},2})$.
+
+The diagram roughly illustrates the interplay of the two kimchi instances.
+
+![Overview](./pickles_structure_overview.svg)
+
+
+The small remark that Step circuit may repeat items 2-3 has to do with the following. In the case where Step 2 consumes several Wrap 1.X (e.g. Wrap 1.1, Wrap 1.2, etc), it must perform all these main Wrap 1.X checks, but also all the deferred Step 1.X checks where Wrap 1.X wraps exactly Step 1.X.
```
Comment (mrmr1993) [inline book/src/pickles/overview.md:23]:
```suggestion
We note that the Step circuit may repeat items 2-3 to handle the following case: when Step 2 consumes several Wrap 1.X (e.g. Wrap 1.1, Wrap 1.2, etc), it must perform all these main Wrap 1.X checks, but also all the deferred Step 1.X checks where Wrap 1.X wraps exactly Step 1.X.
```

File: book/src/pickles/overview.md [hunk 5]
```diff
@@ -1,14 +1,30 @@
...
+
+
+#### On Accumulators
+
+The accumulator is an abstraction introduced for the purpose of this diagram. In practice, each kimchi proof consists of (1) commitments to polynomials, (2) evaluations of them, (3) and the opening proof. What we refer to as accumulator here is actually the commitment inside the opening proof. It is called `sg` in the implementation and is semantically a polynomial commitment to `h(X)` --- the poly-sized polynomial that is built from IPA challenges. It's a very important polynomial -- it can be evaluated in log time, but the commitment verification takes poly time, so the fact that `sg` is a commitment to `h(X)` is never proven inside the circuit. For more details, see [Proof-Carrying Data from Accumulation Schemes](https://eprint.iacr.org/2020/499.pdf), Appendix A.2, where `sg` is called `U`.
```
Comment (mrmr1993) [inline book/src/pickles/overview.md:28]: I think `h(X)` is called `b(X)` in the code (see e.g. `b_poly` and friends). Obviously both names are bad :)

**Review by mrmr1993 — CHANGES_REQUESTED** [review]
Comments about diagrams:
* `MFNS (digest?) -- ignored/bound by SE`
  - `(digest?) is false
* in general, the MFNS names seem unhelpful, I would prefer that they were expanded
* `run application logic of previous statements` should be `run application logic of statement`
* `Assert equal ?????` should be `computed from unpacked preimage when verifying`

File: book/src/pickles/overview.md (hunk 5 above)
Comment (volhovm) [reply book/src/pickles/overview.md:28]: Unfortunately, you're right (and I only saw it now). It's super confusing given that `b` is an evaluation of `b_poly` on the initial `b` coefficients... I'll note it here.

**Comment by volhovm** [comment]
@mrmr1993 ready for another review.

**Review by mrmr1993 — APPROVED** [review]

**Pull request merged**